shell/renderer/electron_renderer_client.cc

```diff
@@ -144,8 +145,8 @@ void ElectronRendererClient::WillReleaseScriptContext(
 
   node::FreeEnvironment(env);
   if (node_bindings_->uv_env() == nullptr) {
```

Comment:
Given now we create isolate data for every environment, we should destruct the corresponding isolate data to avoid leaks. `node_bindings_->uv_env() == nullptr` check is not needed.

```
 node::FreeIsolateData(node_bindings_->isolate_data(context));
 node_bindings_->clear_isolate_data(context);
```

Reply:
Isn't that because you are accessing the environment after it has been destroyed,

```
node::FreeEnvironment(env);
  if (node_bindings_->uv_env() == nullptr) {
    node::FreeIsolateData(node_bindings_->isolate_data(env->context()));
    node_bindings_->clear_isolate_data(env->context());
  }
```

should instead be

```
v8::Local<v8::Context> context = env->context();
node::FreeEnvironment(env);
node::FreeIsolateData(node_bindings_->isolate_data(context));
node_bindings_->clear_isolate_data(context);
```

Reply:
i'm clearly not firing on all cylinders this morning 🙃  corrected.

Reply:
I initially thought so too, yes, but if i do that i see the following:

<details><summary>Details</summary>
<p>

```
#
# Fatal error in ../../v8/src/objects/tagged-impl.h, line 142
# Debug check failed: kCanBeWeak || (!IsSmi() == HAS_STRONG_HEAP_OBJECT_TAG(ptr_)).
#
#
#
#FailureMessage Object: 0x16dae0b580   Electron Framework                  0x000000011c4c8a30 base::debug::CollectStackTrace(void**, unsigned long) + 28
1   Electron Framework                  0x000000011c4b9824 base::debug::StackTrace::StackTrace() + 24
2   Electron Framework                  0x000000011dda2978 gin::(anonymous namespace)::PrintStackTrace() + 40
3   Electron Framework                  0x000000011db63bb4 V8_Fatal(char const*, int, char const*, ...) + 268
4   Electron Framework                  0x000000011db63858 std::__Cr::enable_if<!std::is_function<std::__Cr::remove_pointer<unsigned char>::type>::value && !std::is_enum<unsigned char>::value && has_output_operator<unsigned char, v8::base::CheckMessageStream>::value, std::__Cr::basic_string<char, std::__Cr::char_traits<char>, std::__Cr::allocator<char>>>::type v8::base::PrintCheckOperand<unsigned char>(unsigned char) + 0
5   Electron Framework                  0x00000001191d3a98 v8::Utils::OpenHandle(v8::Context const*, bool) + 156
6   Electron Framework                  0x00000001191d6d14 v8::Context::GetNumberOfEmbedderDataFields() + 28
7   Electron Framework                  0x0000000117d8fff0 electron::NodeBindings::isolate_data(v8::Local<v8::Context>) const + 28
8   Electron Framework                  0x0000000117d95960 electron::WebWorkerObserver::~WebWorkerObserver() + 136
9   Electron Framework                  0x0000000117d95808 electron::WebWorkerObserver::Create() + 104
10  Electron Framework                  0x0000000117d90124 electron::ElectronRendererClient::WorkerScriptReadyForEvaluationOnWorkerThread(v8::Local<v8::Context>) + 88
11  Electron Framework                  0x000000011e02a17c blink::WorkerOrWorkletScriptController::PrepareForEvaluation() + 364
12  Electron Framework                  0x000000011e0296d4 blink::WorkerOrWorkletScriptController::Initialize(blink::KURL const&) + 1156
13  Electron Framework                  0x00000001203b61bc blink::PaintWorkletGlobalScope::Create(blink::LocalFrame*, std::__Cr::unique_ptr<blink::GlobalScopeCreationParams, std::__Cr::default_delete<blink::GlobalScopeCreationParams>>, blink::WorkerReportingProxy&) + 104
14  Electron Framework                  0x00000001203b8b34 blink::PaintWorkletGlobalScopeProxy::PaintWorkletGlobalScopeProxy(blink::LocalFrame*, blink::WorkletModuleResponsesMap*, unsigned long) + 1116
15  Electron Framework                  0x00000001203b499c blink::PaintWorklet::CreateGlobalScope() + 292
16  Electron Framework                  0x000000011effcc80 blink::Worklet::FetchAndInvokeScript(blink::KURL const&, WTF::String const&, blink::WorkletPendingTasks*) + 344
17  Electron Framework                  0x000000011effe094 base::internal::Invoker<base::internal::BindState<void (blink::Worklet::*)(blink::KURL const&, WTF::String const&, blink::WorkletPendingTasks*), cppgc::internal::BasicPersistent<blink::Worklet, cppgc::internal::StrongPersistentPolicy, cppgc::internal::IgnoreLocationPolicy, cppgc::internal::DisabledCheckingPolicy>, blink::KURL, blink::V8RequestCredentials, cppgc::internal::BasicPersistent<blink::WorkletPendingTasks, cppgc::internal::StrongPersistentPolicy, cppgc::internal::IgnoreLocationPolicy, cppgc::internal::DisabledCheckingPolicy>>, void ()>::RunOnce(base::internal::BindStateBase*) + 116
18  Electron Framework                  0x000000011aee234c WTF::ThreadCheckingCallbackWrapper<base::OnceCallback<void ()>, void ()>::Run() + 80
19  Electron Framework                  0x000000011c44f704 base::TaskAnnotator::RunTaskImpl(base::PendingTask&) + 308
20  Electron Framework                  0x000000011c476e64 base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWorkImpl(base::LazyNow*) + 1164
21  Electron Framework                  0x000000011c476494 base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::DoWork() + 104
22  Electron Framework                  0x000000011c40aa28 base::MessagePumpDefault::Run(base::MessagePump::Delegate*) + 152
23  Electron Framework                  0x000000011c477cb8 base::sequence_manager::internal::ThreadControllerWithMessagePumpImpl::Run(bool, base::TimeDelta) + 472
24  Electron Framework                  0x000000011c430d90 base::RunLoop::Run(base::Location const&) + 396
25  Electron Framework                  0x0000000120ac04c4 content::RendererMain(content::MainFunctionParams) + 1408
26  Electron Framework                  0x0000000117f112d0 content::RunOtherNamedProcessTypeMain(std::__Cr::basic_string<char, std::__Cr::char_traits<char>, std::__Cr::allocator<char>> const&, content::MainFunctionParams, content::ContentMainDelegate*) + 696
27  Electron Framework                  0x0000000117f12030 content::ContentMainRunnerImpl::Run() + 724
28  Electron Framework                  0x0000000117f10584 content::RunContentProcess(content::ContentMainParams, content::ContentMainRunner*) + 1228
29  Electron Framework                  0x0000000117f107e8 content::ContentMain(content::ContentMainParams) + 112
30  Electron Framework                  0x0000000117bbed9c ElectronMain + 128
31  Electron Helper (Renderer)          0x000000010231ca00 main + 224
32  dyld                                0x000000019598bf28 start + 2236
Renderer process crashed - see https://www.electronjs.org/docs/tutorial/application-debugging for potential debugging information.
```

</p>
</details> 

Which i'm looking at right now.

Reply:
Try moving the contents of `~WebWorkerObserver` to `WebWorkerObserver::ContextWillDestroy` to avoid using context after it is signaled for destruction.

Reply:
I also tried that in an earlier iteration of this PR - that causes a uv handle closing crash.

Reply:
Addressed in https://github.com/electron/electron/pull/38754/commits/0337db2408b5dde7fae5619bed3b91cbf47fbfce